File: assets/ws/generate-cs-client/sample/src/template.api.wsclient/src/lib/ws/WSServiceBase.cs
using System.Threading.Tasks;

namespace nex.ws
{
    public abstract class WSServiceBase
    {
        #region [ abstract ]
        public abstract string Name { get; }
        #endregion

        #region [ private ]
        private RestClient _rest { get; }
        private HubClient _hub { get; }
        #endregion

        #region [ constructor ]
        public WSServiceBase(RestClient rest, HubClient hub)
        {
            _rest = rest;
            _hub = hub;
        }
        #endregion

        #region [ protected ]
        protected Task<T> Request<T>(string method, object data, object credentials, int timeout = 0)
        {
            return _rest.RequestAsync<T>(new RestRequest
            {
                service = Name,
                method = method,
                data = data,
                credentials = credentials
            }, timeout);
        }
        protected Task Request(string method, object data, object credentials, int timeout = 0)
        {
            return _rest.RequestAsync(new RestRequest
            {
                service = Name,
                method = method,
                data = data,
                credentials = credentials
            }, timeout);
        }

        protected HubNotification<TCredentials, TData> NewEvent<TCredentials, TData>(string eventName){
            return new HubNotification<TCredentials, TData>(_hub, Name, eventName);
        }
        #endregion
    }
}

File: assets/ws/generate-ts-client/demo/lib/WSServiceBase.ts
import { RestClient } from './rest/RestClient';
import { HubClient } from './hub/HubClient';
import { HubNotification } from './hub/HubNotification';

export abstract class WSServiceBase {
    //#region [ abstract ]
    public abstract name: string;
    //#endregion

    //#region [ private ]
    private readonly _rest: RestClient;
    private readonly _hub: HubClient;
    //#endregion

    //#region [ constructor ]
    constructor(rest: RestClient, hub: HubClient) {
        this._rest = rest;
        this._hub = hub;
    }
    //#endregion

    //#region  [ protected ]
    protected request<T = void>(method: string, data: any, credentials: any, timeout?: number) {
        return this._rest.requestAsync<T>(
            {
                service: this.name,
                method,
                data,
                credentials,
            },
            timeout,
        );
    }
    protected newEvent<TCredentials, TData>(event: string) {
        return new HubNotification<TCredentials, TData>(this._hub, this.name, event);
    }
    //#endregion
}
